### PR TITLE
[inductor] Make sure unfuse_addmm and addmm patterns don't overlap

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110235
* #110232

Inductor has two opposing patterns,
```
addmm -> add + mm
add + mm -> addmm
```

This uses the `extra_check` to disable the addmm fusion pattern when the
heuristic to unfuse add is met, for consistency.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_pattern_matcher.py [hunk 1]
```diff
@@ -328,28 +328,38 @@ def fn(a, b, c):
...
             torch._dynamo.reset()
             counters.clear()
+            args = (a, b, c)
             e1, e2 = fn(*args)
             a1, a2 = torch.compile(fn)(*args)
             torch.testing.assert_close(a1, e1)
             torch.testing.assert_close(a2, e2)
-            self.assertEqual(counters["inductor"]["pattern_matcher_count"], 2)
-            self.assertEqual(counters["inductor"]["pattern_matcher_nodes"], 4)
+            count, nodes = (2, 4) if should_fuse else (0, 0)
```
Comment (peterbell10) [inline test/inductor/test_pattern_matcher.py:360]: Note that these cases weren't actually fused previously, it's just that the pattern replaced them with a lowering that did `add + mm`.

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -789,20 +758,67 @@ def is_pointwise_use(use):
...
+    inp = match.kwargs["inp"]
+
+    if not isinstance(inp, torch.fx.Node):
+        return False  # Input is a number
+
+    in_shape = inp.meta["val"].shape
+    matched = len(in_shape) <= 2
+    mm_shape = mat1.meta["val"].shape[0], mat2.meta["val"].shape[1]
+    for i, m in zip(in_shape, mm_shape):
+        matched &= i == 1 or i == m
```
Comment (lezcano) [inline torch/_inductor/fx_passes/post_grad.py:794]:
```suggestion
    matched = len(in_shape) <= 2 and all(i in (1, m) for i, m in zip(in_shape, mm_shape))
```

Now, what happens here if `in_shape` is size 1? You are comparing it just against shape[0]. Shouldn't you reverse it and zip it against `reverse(mm_shape)`?

**Review by lezcano — COMMENTED** [review]
Fair enough. I think the code has a preexisting issue that we should fix tho.

**Review by lezcano — APPROVED** [review]
Even better

File: torch/_inductor/fx_passes/post_grad.py [hunk 3]
```diff
@@ -443,37 +443,6 @@ def cat_slice_cat(match, cat_input, size, dim=1):
...
-)
-def addmm(match, mat1, mat2, inp):
-    if isinstance(inp, ir.TensorBox):
-        inp_shape = inp.get_size()
-        matched = len(inp_shape) <= 2
-        mm_shape = shape_of_mm(mat1, mat2)
-        for i, m in zip(inp_shape, mm_shape):
-            matched &= i == 1 or i == m
-    else:  # inp is a Number
-        matched = False
```
Comment (eellison) [inline torch/_inductor/fx_passes/post_grad.py:470]: Nice to move this away from graph lowering pattern, this was overdue..

**Review by eellison — APPROVED** [review]
Looks good just one comment about checking for input being a tensor

**Pull request closed without merging**